### PR TITLE
HIP compilation on CI targeting RDNA3

### PR DESCRIPTION
**Pull request by blueridanus**
I've just picked the model from `beautiful_mnist.py` to check if it compiles. Might be worth testing a few more (which ones?)

### CODE REVIEW

**Comment by g1y5x3** [comment]
Can you add the [SpeedyResNet](https://github.com/tinygrad/tinygrad/blob/master/examples/hlb_cifar10.py#L66) with both single and half precion? The half precision is suppose to fail this test.

File: test/test_hip_rdna3.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class TestHIPCompilationRDNA(unittest.TestCase):
+  def test_compile_hip(self):
+    if not getenv("HIP"): raise unittest.SkipTest("testing HIP->rdna3 compilation needs HIP=1")
+
+    model = Model()
+
+    input = Tensor.rand(512,1,28,28, device="CPU").numpy()
+    result = model(Tensor(input))
```
Comment (g1y5x3) [inline test/test_hip_rdna3.py:20]: can't you just use a tinygrad tensor instead of converting it numpy and then convert back?

Comment (blueridanus) [reply test/test_hip_rdna3.py:20]: oops. forgot it there from a poor attempt at fixing it failing on the rand op (the right was to filter off LoadOps above)

**Comment by blueridanus** [comment]
> Can you add the [SpeedyResNet](https://github.com/tinygrad/tinygrad/blob/master/examples/hlb_cifar10.py#L66) with both single and half precion? The half precision is suppose to fail this test.

Done. I've marked the half precision text as xfail.

**Comment by g1y5x3** [comment]
based on the half precision error that I know, it was reported from `hipCheckStatus` that's inside `hip.hiprtcCompileProgram` followed with error message from the compiler like 
```
/tmp/comgr-d9b0aa/input/<null>:34:238: note: built-in candidate operator-(float, float)
  vstore_half4(make_float4((((((val0).x*4.0f)+15.0f)<cast0)!=0?(cast0-1.0f):cast0),(((((val0).y*4.0f)+15.0f)<cast1)!=0?(cast1-1.0f):cast1),(((((val0).z*4.0f)+15.0f)<cast2)!=0?(cast2-1.0f):cast2),(((((val0).w*4.0f)+15.0f)<cast3)!=0?(cast3-1.0f):cast3)), 0, data0+(gidx0*16)+(lidx1*4));
                                                                                                                                                                                                                                             ^
/tmp/comgr-d9b0aa/input/<null>:34:238: note: built-in candidate operator-(double, float)
...
8 errors generated when compiling for gfx1100.
```
Is it possible to get that feedback from CI?

**Comment by g1y5x3** [comment]
Ah, try `unittest.main(verbosity=2)` in `__main__`

File: test/test_hip_rdna3.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+from tinygrad.ops import LoadOps
+from tinygrad.codegen.kernel import LinearizerOptions
+from tinygrad.codegen.linearizer import Linearizer
+from tinygrad.renderer.hip import HIPRenderer
+from examples.beautiful_mnist import Model as MNIST
+from examples.hlb_cifar10 import SpeedyResNet
+import extra.hip_wrapper as hip
+
+
+def check_rdna3_compilation(result:Tensor):
```
Comment (geohot) [inline test/test_hip_rdna3.py:14]: This should not be copied code, it should use all the tinygrad stuff, just be okay with not running the kernel

File: tinygrad/runtime/ops_hip.py [hunk 3]
```diff
@@ -27,9 +27,9 @@ def _cached_bufkey(self, size, dtype, device): return (device, size*dtype.itemsi
 class _HIP:
   def __init__(self, device=None):
     self.default_device = device or getenv("HIP_DEFAULT_DEVICE")
-    hip.hipSetDevice(self.default_device)
-    self.device_count = hip.hipGetDeviceCount()
-    self.allocator = HIPAllocator(hip.hipGetDeviceProperties(self.default_device).totalGlobalMem)
+    self.device_count = 0 if getenv("CI") else hip.hipGetDeviceCount()
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:30]: from tinygrad.helpers import CI

File: test/test_hip_rdna3.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import unittest
+from tinygrad import Tensor
+from tinygrad.helpers import dtypes, getenv
+from examples.beautiful_mnist import Model as MNIST
+from examples.hlb_cifar10 import SpeedyResNet
+
+
+class TestHIPCompilationRDNA(unittest.TestCase):
+  def test_compile_hip_mnist(self):
+    if not getenv("HIP"): raise unittest.SkipTest("testing HIP->rdna3 compilation needs HIP=1")
```
Comment (geohot) [inline test/test_hip_rdna3.py:11]: Device.DEFAULT =  "HIP"

**Comment by geohot** [comment]
Congrats! Can pay bounty via PayPal or USDC, e-mail george@tinygrad.org to claim

**Pull request merged**